evalml/preprocessing/data_splitters/smote_split.py

```diff
@@ -0,0 +1,124 @@
+
+            Arguments:
+                X (ww.DataTable): DataTable of points to split
+                y (ww.DataTable): DataColumn of points to split
+
+            Returns:
+                tuple(ww.DataTable, ww.DataColumn): A tuple containing the resulting X and y post-transformation
+        """
+        X_pd, y_pd = self._to_woodwork(X, y)
+        X_transformed, y_transformed = self.kmsmote.fit_resample(X_pd, y_pd)
+        return (_convert_to_woodwork_structure(X_transformed), _convert_to_woodwork_structure(y_transformed))
+
+
+class KMeansSMOTECVSplit(StratifiedKFold):
+    """Split the data into KFold cross validation sets and balances the training data using K-Means SMOTE.
+       Keeps the validation data the same. Works only on continuous, numeric data."""
+
+    def __init__(self, sampling_strategy='auto', k_neighbors=2, n_splits=3, shuffle=True, random_state=0, **kwargs):
+        super().__init__(n_splits=n_splits, shuffle=shuffle, random_state=random_state)
+        error_msg = "imbalanced-learn is not installed. Please install using 'pip install imbalanced-learn'"
+        im = import_or_raise("imblearn.over_sampling", error_msg=error_msg)
+        self.kmsmote = im.KMeansSMOTE(sampling_strategy=sampling_strategy, k_neighbors=k_neighbors, random_state=random_state, **kwargs)
+        self.random_state = random_state
+        self.n_splits = n_splits
+
+    def _to_woodwork(self, X, y, to_pandas=True):
+        """Convert the data to woodwork datatype"""
+        X_ww = _convert_to_woodwork_structure(X)
+        if not is_all_numeric(X_ww):
+            raise ValueError('Values not all numeric or there are null values provided in the dataset')
+        y_ww = _convert_to_woodwork_structure(y)
+        if to_pandas:
+            X_ww = _convert_woodwork_types_wrapper(X_ww.to_dataframe())
+            y_ww = _convert_woodwork_types_wrapper(y_ww.to_series())
+        return X_ww, y_ww
+
+    def split(self, X, y=None):
+        """Divides the data into training and testing sets
+
+            Arguments:
+                X (ww.DataTable): DataTable of points to split
+                y (ww.DataTable): DataColumn of points to split
+
+            Returns:
+                tuple(list): A tuple containing the resulting X_train, X_valid, y_train, y_valid data.
+        """
+        X, y = self._to_woodwork(X, y)
+        for i, (train_indices, test_indices) in enumerate(super().split(X, y)):
+            X_train, X_test, y_train, y_test = X.iloc[train_indices], X.iloc[test_indices], y.iloc[train_indices], y.iloc[test_indices]
+            X_train_resample, y_train_resample = self.kmsmote.fit_resample(X_train, y_train)
+            X_train_resample, y_train_resample = self._to_woodwork(X_train_resample, y_train_resample, to_pandas=False)
+            X_test, y_test = self._to_woodwork(X_test, y_test, to_pandas=False)
+            yield iter(((X_train_resample, y_train_resample), (X_test, y_test)))
```

Comment:
I think same thing here, it might be worth having like a BaseCVSplit class for this functionality.